### PR TITLE
[DATA-147] Extract pointcloud and implement grpc endpoint

### PR DESCRIPTION
**Pull request by kkufieta**
https://viam.atlassian.net/browse/DATA-147

Implement the pointcloud grpc endpoint. Tested on:
* Raspberry Pi OS Lite (64-bit)
* 2.4 GHz 8-Core Intel Core i9; macOS Monterey

In offline mode.

### CODE REVIEW

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc [hunk 1]
```diff
@@ -56,11 +56,16 @@ ::grpc::Status SLAMServiceImpl::GetMap(ServerContext *context,
...
-        return grpc::Status(
-            grpc::StatusCode::UNIMPLEMENTED,
-            "GetMap for mime_type \"pointcloud/pcd\" is not yet implemented.");
+        std::stringbuf buffer;
+        bool has_points = ExtractPointCloudToBuffer(buffer);
+        if (has_points) {
+            common::v1::PointCloudObject *pco = response->mutable_point_cloud();
+            pco->set_point_cloud(buffer.str());
+        } else {
+            LOG(ERROR) << "map pointcloud does not have points yet\n";
```
Comment (tessavitabile) [inline slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:65]: [nit] I think we don't need `\n` at the end of log lines.

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc [hunk 2]
```diff
@@ -218,6 +223,69 @@ std::string SLAMServiceImpl::PaintMap() {
...
+        << "SIZE 4 4 4 4\n"
+        << "TYPE F F F I\n"
+        << "COUNT 1 1 1 1\n"
+        << "WIDTH " << number_points << "\n"
+        << "HEIGHT " << 1 << "\n"
+        << "VIEWPOINT 0 0 0 1 0 0 0\n"
+        << "POINTS " << number_points << "\n"
+        << "DATA binary\n";
+
+    for (auto trajectory_node : trajectory_nodes) {
```
Comment (tessavitabile) [inline slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:258]: [nit] `const auto&&`

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc [hunk 3]
```diff
@@ -218,6 +223,69 @@ std::string SLAMServiceImpl::PaintMap() {
...
+        cartographer::sensor::PointCloud local_gravity_aligned_point_cloud =
+            trajectory_node.data.constant_data
+                ->filtered_gravity_aligned_point_cloud;
+
+        // We're only applying the transformation of the `global_pose` on the
```
Comment (tessavitabile) [inline slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:263]: Is "transformation" supposed to be "translation" here?

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc [hunk 4]
```diff
@@ -218,6 +223,69 @@ std::string SLAMServiceImpl::PaintMap() {
...
+    std::ostream oss(&buffer);
+    bool has_points = false;
+
+    cartographer::mapping::MapById<cartographer::mapping::NodeId,
+                                   cartographer::mapping::TrajectoryNode>
+        trajectory_nodes;
+    {
+        std::lock_guard<std::mutex> lk(map_builder_mutex);
+        trajectory_nodes =
+            map_builder.map_builder_->pose_graph()->GetTrajectoryNodes();
```
Comment (tessavitabile) [inline slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:236]: Once we extract `trajectory_nodes` is it possible for them to be modified by `map_builder`? I want to check that accesses to `trajectory_nodes` are safe after we release the mutex.

**Review by tessavitabile — COMMENTED** [review]
Nice! A few nits and one question about concurrency.

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc (hunk 1 above)
Comment (kkufieta) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:65]: right, forgot about that!

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc [hunk 5]
```diff
@@ -56,11 +56,16 @@ ::grpc::Status SLAMServiceImpl::GetMap(ServerContext *context,
...
-            grpc::StatusCode::UNIMPLEMENTED,
-            "GetMap for mime_type \"pointcloud/pcd\" is not yet implemented.");
+        std::stringbuf buffer;
+        bool has_points = ExtractPointCloudToBuffer(buffer);
+        if (has_points) {
+            common::v1::PointCloudObject *pco = response->mutable_point_cloud();
+            pco->set_point_cloud(buffer.str());
+        } else {
+            LOG(ERROR) << "map pointcloud does not have points yet\n";
+            return grpc::Status(grpc::StatusCode::UNIMPLEMENTED,
```
[the remark this answers is not in the document]
Comment (kkufieta) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:66]: ooh good catch

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc [hunk 6]
```diff
@@ -218,6 +223,69 @@ std::string SLAMServiceImpl::PaintMap() {
...
+                                   cartographer::mapping::TrajectoryNode>
+        trajectory_nodes;
+    {
+        std::lock_guard<std::mutex> lk(map_builder_mutex);
+        trajectory_nodes =
+            map_builder.map_builder_->pose_graph()->GetTrajectoryNodes();
+    }
+
+    long number_points = 0;
+    for (auto trajectory_node : trajectory_nodes) {
```
[the remark this answers is not in the document]
Comment (kkufieta) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:240]: oh forgot this, too, thanks for the reminder and catching this

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc (hunk 3 above)
Comment (kkufieta) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:263]: yes, thanks for catching this!

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc [hunk 7]
```diff
@@ -218,6 +223,69 @@ std::string SLAMServiceImpl::PaintMap() {
...
+        // world frame is already taken care of.
+        cartographer::sensor::PointCloud global_point_cloud =
+            cartographer::sensor::TransformPointCloud(
+                local_gravity_aligned_point_cloud,
+                cartographer::transform::Rigid3f(
+                    trajectory_node.data.global_pose.cast<float>()
+                        .translation(),
+                    cartographer::transform::Rigid3f::Quaternion::Identity()));
+
+        for (auto point : global_point_cloud) {
```
[the remark this answers is not in the document]
Comment (kkufieta) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc]: Does actually not compile with `const`, setting it to be `auto&& point`

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc (hunk 4 above)
Comment (kkufieta) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:236]: Looked into it together, and looks to be a safe copy including one shared pointer

**Review by tessavitabile — APPROVED** [review]

**Pull request merged**